src/test/java/uk/gov/dwp/dataworks/DataKeyServiceApplicationTests.java

```diff
@@ -1,15 +1,126 @@
-@SpringBootTest
+@SpringBootTest()
+@ActiveProfiles("Test")
+@AutoConfigureMockMvc
 public class DataKeyServiceApplicationTests {
 
+    @Autowired
+    private CurrentKeyIdProvider currentKeyIdProvider;
+
+    @Autowired
+    private DataKeyGeneratorProvider dataKeyGeneratorProvider;
+
+    @Autowired
+    private DataKeyDecryptionProvider dataKeyDecryptionProvider;
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Before
+    public void setup() {
+        Mockito.reset(currentKeyIdProvider, dataKeyGeneratorProvider, dataKeyDecryptionProvider);
+    }
+
     @Test
     public void contextLoads() {
     }
+
+    @Test
+    public void pingWorks() throws Exception {
+        mockMvc.perform(get("/ping"))
+                .andExpect(status().isOk())
+                .andExpect(content().string(""));
+    }
+
+    @Test
+    public void generateDataKeyRequest() throws Exception {
+        given(currentKeyIdProvider.getKeyId()).willReturn("myKeyId");
+
+        GenerateDataKeyResponse response = new GenerateDataKeyResponse("encryptionId", "plainKey", "cipherKey");
+
+        given(dataKeyGeneratorProvider.generateDataKey("myKeyId"))
+                .willReturn(response);
+
+        mockMvc.perform(get("/datakey"))
+                .andExpect(status().isCreated())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+
+        // Must also allow a trailing slash
+        mockMvc.perform(get("/datakey/"))
+                .andExpect(status().isCreated())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+    }
+
+    @Test
+    public void decryptRequest() throws Exception {
+        DecryptDataKeyResponse response = new DecryptDataKeyResponse("decryptKeyId", "iv", "plaintextDataKey");
+        given(dataKeyDecryptionProvider.decryptDataKey("encryptionKeyId", "cipher text"))
+                .willReturn(response);
+
+        mockMvc.perform(post("/datakey/actions/decrypt?keyId={keyId}", "encryptionKeyId").content("cipher text"))
+                .andExpect(status().isOk())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+    }
+
+    @Test
+    public void currentKeyIdProviderThrowsError() throws Exception {
+        given(currentKeyIdProvider.getKeyId()).willThrow(new CurrentKeyIdFailure());
+
+        mockMvc.perform(get("/datakey"))
+                .andExpect(status().isServiceUnavailable());
+    }
+
+    @Test
+    public void dataKeyGeneratorProviderThrowsError() throws Exception {
+        given(dataKeyGeneratorProvider.generateDataKey(any())).willThrow(new DataKeyGenerationFailure());
+
+        mockMvc.perform(get("/datakey"))
+                .andExpect(status().isServiceUnavailable());
+    }
+
+    @Test
+    public void setDataKeyDecryptionProviderThrowsError() throws Exception {
+        given(dataKeyDecryptionProvider.decryptDataKey(any(), any())).willThrow(new DataKeyDecryptionFailure());
```

Comment:
Would suggest making the mock as specific as possible without any() so that the test is verifying that the specific case within it has caused the exception to be thrown - rather than the result of some unexpected side affect.